release-content/0.15/release-notes/14791_Required_Components.md

```diff
@@ -1,4 +1,421 @@
-<!-- Required Components -->
-<!-- https://github.com/bevyengine/bevy/pull/14791 -->
+First: buckle up because **Required Components** is one of the most profound improvements to the Bevy API surface since Bevy was first released.
```

Comment:
Starting a thread to discuss my comment above, in the interest of making this discussion manageable.

Reply:
I think I'm sold: this is a compelling defense. I'll take a pass to see if we can help it come through clearly in the blog post itself, and make the "flexibility must be designed" point well.

Reply:
Thanks for the considered write-up @cart!

The schools of thought you have outlined seem a little unclear to me. For example, I would say that I consider components to be just data, but that behaviours are governed by the set of components on an entity. I would not say that I want to remove required components, rather that if we encourage library vendors to use required components for convenience, that they're going to seriously compromise the composibility inherent in Bevy's ECS.

I'm also not totally sure about the requirements you seem to be outlining.

You say that library/app authors shouldn't "mutate existing upstream concepts/assumptions" - but this is very vague. Obviously you shouldn't be modifying someone elses assumptions, but if it's something that library and app developers _can_ modify then the mistake was upstream: they made an erroneous assumption.

A key example of this is the claim that if required components don't exist, libraries will be unable to maintain their invariants. Could you provide an example of an invariant that required components help to uphold?

As far as I can tell, components are always accessed optionally: you either directly call `get` which returns an `Option<>`, or you query for the component, which will only yield entities that match. The only edge-case I can think of where the invariants of a library could become invalidated is if they provide _different_ queries to systems and _expect them to act on the same set of entitites_. For example, having a cleanup system _with a more specific query_, meaning it doesn't run for some entities and leaks. However, I would consider this a very clear cut library bug, which is very straight-forward to address.

Another thing you raise is "re-using components for different purposes". I would argue that in both cases, the specific semantic meaning of the component **is** important (and it's up to the consumer to not abuse that, or rather, if they try and use it for a different meaning, it should be clear that other systems might take it as the original meaning). However, if you do need to have entities with the same semantic aspect, but aren't planning to fulfil the contract of existing systems, I think it's desirable to be able to re-use the component (if people write their own render pipelines, say for point clouds, I would expect them to integrate with `Visibility` in the same way that meshes & meshlets do).

I agree with your argument that building generic components is inherently more complicated, and I don't think that's something Bevy should be trying to do. However again, it feels like a non-sequitur.

It makes sense to me that you would _require_ private components you need to maintain your invariants (which would be invisible to consumers of your library), but potentially _include_ components which are part of your contract for better user experience (in lieu of an editor or authoring flow).

As we've discussed before, `Transform` is an excellent example that I think should not be required but should be included. It seems intuitive to me that the rendering modules should only care about `GlobalTransform`.

Reply:
> they're going to seriously compromise the composibility inherent in Bevy's ECS.

The whole angle of my argument is that composability is not _actually_ inherent in the ways that matter. Composability is something that is carefully created. Required Components don't get in the way of that. In fact, they help facilitate it by making the contract clearer.

> Could you provide an example of an invariant that required components help to uphold?

My Visibility examples cover a variety of these. Things as simple as "All Players have Health" or "Health requires Character" also clearly fall into this category. Being able to rely on a conceptual "whole" composed by many parts is critical.

> but aren't planning to fulfil the contract of existing systems,

This is part of the problem. The systems are just one part of the contract. How all the pieces relate to each other is the contract. And in most cases, the pieces have not been carefully designed to be consumed piecemeal at the arbitrary whims of downstream developers.

> As we've discussed before, Transform is an excellent example that I think should not be required but should be included. It seems intuitive to me that the rendering modules should only care about GlobalTransform.

This does largely make sense to me. I think there is a reasonable case to be made for Transform to be "included" and not "required". But I think that is largely an exception to the rule, and `GlobalTransform` is a clear case of being a well-defined largely "general purpose" concept. The "optional and included for convenience"  category of component is something that will come up. I'm not fighting that. Only stating that (1) `require` should be the default and the focus (2) the usage of `include` requires careful thought.

Reply:
> The whole angle of my argument is that composability is not actually inherent in the ways that matter.

In traditional ECSes, you do often get composability "for free" because systems only interact with components that are part of their contract (or their own private components). Often it's left as an exercise to the consumer to make sure that components required for follow-on behaviour are present (if they want that behaviour). Conversely, they are free to introduce their own behaviours that write to inputs or read from outputs of other systems.

> My Visibility examples cover a variety of these.

You have provided examples of requirements, and if we assume that required components cannot be removed, you've provided invariants, but it's not clear to me that anything would depend on them. In those examples, what would be relying on the invariants?

> How all the pieces relate to each other is the contract.

You need to be more specific here. Components can't relate to each other inherently as they're just data. Is your point here that the contract is that these components only appear together in archetypes?

Required components limit the archetypes that are allowed to exist. Queries already exist to limit the archetypes that systems act upon. It's already possible to ensure the behaviours you're referring to only operate on entities that _do_ have all the components in question. So why is it important to disallow archetypes that only contain some of the components?

It seems to me that the same result could be achieved by using additional `With<>` clauses in systems to depend on their downstream components, and including components by default rather than requiring them.

For the goal of improving new user experience, I think this has the same upsides. For the Bevy developers, I think it helps prevent accidentally creating behaviour people depend on. For developers, it seems less restrictive than required components.

Reply:
(I'm replying up here but this is in relation to the thread of the topic about physics.)

@Jondolf there's an important nuance I'm worried you've missed here: I don't think that the dichotomy is between hard requirements and soft requirements.

From my point of view, both scenarios are offering hard requirements. An example of soft requirements would be automatically assuming defaults if the components were missing.

I think that include-by-default, with suitably strict queries provides both the hard requirements at execution, the removal of some foot-guns for new users, **and** allows flexibility without undermining invariants.

Reply:
> In traditional ECSes, you do often get composability "for free" because systems only interact with components that are part of their contract 

I fully acknowledge that traditional ECS helps facilitate composability in this way (we all know how ECS works here). My intent when I said "is not actually inherent _in the ways that matter_", was to express that this free-form mixing of contexts that ECS allows does not inherently solve the hard problems of building composable APIs (predictable behaviors, correctness, well defined scope, preventing people from stepping on each other's toes, etc).

> but it's not clear to me that anything would depend on them. In those examples, what would be relying on the invariants?

Allowing someone to strip away the ViewVisibility / InheritedVisibility components is a signal to the user that stripping them out and using `Visibility` standalone is supported, outside of a hierarchical or view context. That is not the case for the current implementation (for the reasons I stated). Those are the "invariants". I think it is important to call out that these "invariants" will show up by nature of building a specific thing in a specific context (ex: assuming that propagation happens only when all of the components are present in each member of the hierarchy). Many of them will be choices that were not consciously made.

> How all the pieces relate to each other is the contract. Is your point here that the contract is that these components only appear together in archetypes?

Yes that is one of my points. It is a statement that "these components exist to _collectively_ produce a set of behaviors, they are not designed to be used alone, and the implications of doing so have either not been explored or have been rejected". Restricting the archetype is a part of that (and a benefit of that), but it is just one expression of that set of values.

> It seems to me that the same result could be achieved by using additional `With<>` clauses in systems to depend on their downstream components,

This is another one of my points. If `Visibility` means "hierarchical / view visibility" only when the hierarchy and view components are also present, that means consumers needs to express that manually on every query (ex: via `With` clauses) when they only want "hierarchical / view visibility". I'm arguing that it is a simpler / easier to use system if we just let `Visibility` be the "hierarchical / view visibility" component. One piece in the system we've built, rather than a standalone piece of functionality.

"ECS queries using multiple components as filters to selectively enable logic" does not buy us "free composable behavior", as evidenced by `Visibility` silently blocking hierarchy propagation, even in absence of the view / hierarchy components. There is an infinite number of ways that a system could "partially break" when attempting to use a component out of the context it was original built for.

Even if (by happenstance) stripping out components built to work together produces the desired behavior, and even if you manage to layer new behavior on top (ex: `TransdimensionalVisibility` or something), you've still made the weirdness factor significantly higher, by nature of _fundamentally changing_ what it means to interact with `Visibility` in some cases. Then sometimes when you change `Visibility`, it is doing "transdimensional visibility" and sometimes it is doing "hierarchical / view visibility".

There _may_ be an intersection between "transdimensional visibility" and "hierarchical / view visibility". But the point is _the author of `Visibility` did not build it with that scenario in mind or define the set of rules it would follow in that context_. Maybe we _want_ a more general `Visibility` component that can accommodate "transdimensional visibility" (and other such scenarios). To support that, we need to discuss what a component like that would look like, ensure all of the functionality we have built so far is compatible with the scenarios we want to cover, document how it is intended to be used, and then ensure the changes we make going forward don't break those scenarios.

> For the goal of improving new user experience, I think this has the same upside

It does not solve the fundamental issue, which has been stated a number of times already. My goal for the system (and how we talk about it) is to strongly guide people toward building cohesive systems that are easy to use / understand / document. That means _intentional API design_. If someone isn't intentionally baking "generic-ness / optional-ness" into their design, we should not assume it is there.